### ISSUE TITLE
Firespitter / AJE RPM settings integration

### ISSUE THREAD

**Issue by theonegalen** [MOARdV/AvionicsSystems#56]
Feature request: integration with Firespitter and Advanced Jet Engine to allow their RPM (revolutions per minute) settings (available in the right click menu of correctly configured propeller engines) to be controlled via an IVA prop, and so the result can be displayed on another prop. Thank you.

Priority is pretty low, but it would add a nice functionality to my Warbird Cockpit IVAs once I switch over to MAS, which I will expedite once the navigation radio / beacon code is ready.

**Comment by theonegalen** [comment]
Let me know if this is the wrong place for feature requests.

**Comment by MOARdV** [comment]
> Let me know if this is the wrong place for feature requests.

This is definitely the right place.  It is a lot harder for me to forget the request when it's here. :)

I'll look in to Firespitter and AJE to see how readily I can control those settings.

**Comment by theonegalen** [comment]
In the same way, once navigation is set up, how about integrating with Waypoint Manager? Or would it be easier to crosstalk MAS and WM through the stock waypoint / KerbNet function?

**Comment by MOARdV** [comment]
I finally tried out WM earlier this year, and it was handy for marking interesting features on Kerbin (ground stations, KSC2, etc).  I do want to find a way to integrate with it - the main issue is that locations flagged in WM need an extra step to convert to a navigation beacon (set their radio types / callsigns / frequencies).  That would be controlled by a UI from the Space Centre screen, most likely.

**Comment by MOARdV** [comment]
It looks like reporting RPMs from Firespitter (maxRPM and currentRPM) will be easy to do.  They're both easy to access.  From a skim of the Firespitter code, RPM is an informational value.  The engine module updates it every FixedUpdate based on throttle setting and resource usage, so I don't see what purpose writing it will do.  I didn't locate a Firespitter prop that changed RPM (admittedly, it was a cursory look on GitHub, not a scan through downloaded files).

**Comment by theonegalen** [comment]
No, Firespitter itself does not contain a RPM / prop pitch selector. IVA prop. It does have a tachometer.

**Comment by MOARdV** [comment]
Ah, I misunderstood your request, perhaps?  I thought "to be controlled via an IVA prop" indicated wanting to tell the mod what the values are.  For reading, this will be a simple change.

**Comment by theonegalen** [comment]
I do want to be able to modify the variables through an IVA prop, it is just that this prop does not yet exist. Currently, one can only set the values in the right click menu of the Firespitter engines. I am sorry if I confused you.

**Comment by MOARdV** [comment]
Quite all right.  I looked at the code, and it did not appear that RPMs were intended to be modified by the player.  However, I haven't actually used a Firespitter engine in-game to see what it does, so I'm a bit ignorant on the details.

**Comment by theonegalen** [comment]
I'll look at it again. I may be getting it mixed up with Advanced Jet Engine's propeller codd

**Comment by theonegalen** [comment]
Sorry that I never got back to you. So here's the setup: AJE has a right click context menu that allows you to set the RPMs for many of it's propellers, and Firespitter doesn't have that, but does have an internal RPM tachometer.

Would it be possible to make both of them compatible with MAS, allowing engine AJE RPM control via an internal knob or lever prop, and the display of the RPM values from either AJE or Firespitter?

**Comment by MOARdV** [comment]
> Would it be possible to make both of them compatible with MAS, allowing engine AJE RPM control via an internal knob or lever prop, and the display of the RPM values from either AJE or Firespitter?

Yes, I think that would be feasible.  I think the biggest trick will be coming up with a way to identify engines individually and predictably - I currently can access them individually, but I can't guarantee what order they're in. On a 4-engine aircraft, it could conceivably be something like 4-1-2-3 when looking at the wing.  I have an idea how to do it, but it will require the player to take steps to configure the engines in the SPH.

**Comment by theonegalen** [comment]
Sure, some kind of SPH tweakable identifying part module that can be added with a @PART[*]:MODULE[ModuleEnginesAJEPropeller] cfg would seem to do that. Or maybe you'll come up with a more elegant solution. :)

**Comment by MOARdV** [comment]
Okay, here are the available fields in ModuleEnginesAJEPropeller:
* propRPM (read-only)
* propPitch (read-only)
* propThrust (read-only)
* manifoldPressure (read-only)
* chargeAirTemp (read-only)
* netExhaustThrust (read-only)
* netMeredithEffect (read-only)
* brakeShaftPower (read-only)
* boost (read/write)
* rpmLever (read/write)
* mixture (read/write)

I would think some of the read-only fields might not be needed for IVA (exhaust thrust and meredith effect?), but it's relatively easy to make them all available if needed.

**Comment by theonegalen** [comment]
Ok, so I would like access to all of it, if you don't mind. There are some items I won't use right at first, but eventually I'll be making MFD pages, even if most of the less important information will be optional.

Oh man, I'm so excited to get to actually have an effective manifold pressure gauge and tachometer instead of the hacked together nonsense I'm using right now.

I didn't even realize there was a mixture control. This is even more in depth than I thought, and I might need more levers than I asked alexustas for, especially once we get into multi-engined planes. I wish there was more AJE documentation. Regarding multiple engines, since we don't have more than one control axis for engines in ksp

Thank you so very much for this.

**Comment by theonegalen** [comment]
Accidentally hit the wrong button on mobile. Sorry.

As I was saying, since we don't have more than one control axis for engines in ksp, and the AFBW mod which changed that seems extremely dead, I would like to be able to set all the read/write fields for multiple engines together, so I or the player wouldn't have top go through four clickable levers to changes the RPM settings of each engine individually. However, I would then like to be able to change them individually later depending on circumstance. 

Maybe a prop that binds multiple levers together into one would be doable. I guess anything is doable with the scripting in MAS.

**Comment by MOARdV** [comment]
I can augment the existing implementation so you can set all engines with one call, although the 'get' functions will be somewhat nonsensical in that case - they'd either have to return the first engine's results, or they'd have to average the results.

**Comment by MOARdV** [comment]
One thing I don't have yet - some controls, like the prop RPM control, don't affect all engines (as I belatedly realized while testing with a fixed-pitch prop), so the engines that can make use fully of these capabilities is somewhat limited.

**Comment by MOARdV** [comment]
One thing that's missing from the data is the engine temperature.  The solver computes this value, but it does not provide direct access to it - MAS would have to get the string displayed in the UI and convert it back to a number, which seems a bit too convoluted.  I'll put in a request on the SolverEngines mod to see if an accessor can be provided.

**Comment by theonegalen** [comment]
That would be great, thanks!

**Comment by theonegalen** [comment]
Now that I've thought about it, what fields could be available from the AJE jet engine module? It might be nice to get various temperatures, N1, N2, etc, if they are available. Especially since it's possible to burn up your engine so much more easily in AJE/FAR than in stock.

**Comment by MOARdV** [comment]
The jets don't seem to expose much beyond the core ModuleEnginesSolver data (which is current throttle position and a string to display internal temperature, data that can be exposed for propellers, as well).  The only gui data I see in the AJE Jet is compression ratio, core throttle, and afterburner throttle.

**Comment by theonegalen** [comment]
Is the internal temperature given by ModuleEnginesSolver different from that given by the KSP thermal right-click menu?

**Comment by MOARdV** [comment]
> Is the internal temperature given by ModuleEnginesSolver different from that given by the KSP thermal right-click menu?

I assume it is, but I haven't checked.  The internal temperature from ModuleEnginesSolver is displayed in the right-click menu under "Eng. Internal Temp" - although it looks like the propeller engines rename that to "Exhaust temp".

It looks like all of the info on the guts of a jet are private data - AJE knows what the temperature at each stage is, but I have no way to access that data because it's two layers buried in the class behind "private" qualifiers.

**Comment by theonegalen** [comment]
In that case, I would like access to the engine temperature field, please. Maybe core and afterburner throttle. I'm thinking a light or panel that clicks on when the engine goes from MIL to AB.

Thank you very much for all of this work. I doubt anyone but me will use these.

**Comment by MOARdV** [comment]
> Maybe core and afterburner throttle. 

Those should be easy.  Engine Temp will be trickier without some changes in SolverEngines or AJE. I can get it from the string in SolverEngines, but it's not efficient having to convert the string back to a number.  I'll also check to see if the stock part temperature has any correlation to the engine temp.

> Thank you very much for all of this work. I doubt anyone but me will use these.

At one time, DeputyLOL was working on a fairly impressive B737 cockpit, IIRC.  Some of the features here are things he asked for, but I wasn't far enough along on basic MAS functionality to cover these sorts of things.  And even if no one else uses it, it'll make your IVAs that much cooler. :)

**Comment by MOARdV** [comment]
Finally added core / afterburner throttle position queries for AJE jets.  I'll see if any other values can be drawn out to MAS, but it doesn't currently look like there is, unfortunately.

**Comment by MOARdV** [comment]
Okay, I've added current and maximum temperature readings for the jet engine.  Updated version of MAS is on [DropBox](https://www.dropbox.com/s/lgffr6rtdf0cc6b/AvionicsSystems-0.14.0b2.zip?dl=0).  I think I'm going to decline adding support for Firespitter as well, since I don't use either of these mods myself, so I'd like to minimize the support overhead.

Since I've already started with AJE support, I'm willing to add additional interactions with it as needed.

**Comment by theonegalen** [comment]
Fantastic, thank you. And AJE just updated to support rotary wing (helicopter) engines as well, so I'll definitely want some of those controls available too. I just don't know exactly what they are since I haven't been able to fire up KSP for a few weeks due to the end of the semester.

**Comment by MOARdV** [comment]
Informational fields I see for helicopters - ShaftPower (HP), and RPM.  There's a cyclic control field that looks like it is both information (current settling) and a controllable field.  The separate tail rotor module doesn't look like it has any values that would need exposed through MAS (most of what it has repeats the main rotor values).

**Comment by theonegalen** [comment]
RPM would be nice.  Can you read the maximum RPM of an engine as well as it's current RPM or percentage?
HP same, and Cyclic.

Cyclic actually confuses me a little. I might need to download and test the new version of AJE with the helicopter blades before I figure all this out.

Luckily, I just finished my last class of the semester, so I can try that tonight!

**Comment by MOARdV** [comment]
Since the helicopter section is pretty sparse, as far as info and control that MAS can provide, I'm going to go ahead and close this issue.  If there is something else in AJE that MAS should support, please go ahead and open a new issue.
